# 内置函数异常

1. len ,new, make

   ```
   new ： 用来分配内存，主要用来分配值类型，比如int，float32 返回的是指针
   new出来的是指针类型，值是一个内存地址，该内存地址指向的才是一个值
   ```

   ```
   make 用来分配内存，只要分配引用类型，比如channel（管道），map,slice
   ```

   ###  错误处理（异常处理）

2. 关键字 defer，panic，recover

   ```
   func test0125(){
   	//使用defer + recover来处理和捕获异常
   	defer func() {
   		err := recover()//recover是内置函数，可以捕获到异常
   		if err != nil{//err的默认值是nil,如果不为nil，说明捕获到了异常
   			fmt.Println("异常信息： ",err)
   		}
   	}()
   	num1 := 10
   	num2 := 0
   	res := num1/num2
   	fmt.Println("res =",res)
   }
   ```

   3. 自定义错误

    ````
   func main() {
   	err := readconfig("init.confi")
   	if err != nil {
   		//如果读取文件错误，则输出这个错误，并终止程序
   		panic(err)
   	}
   	fmt.Println("之后的代码")
   
   }
   
   //函数读取配置文件init.config文件的信息
   //如果文件名错误，则返回一个自定义错误
   func readconfig(name string) (err error) {
   	if name == "init.config" {
   		return nil
   	} else {
   		//返回一个自定义异常
   		return errors.New("读取的文件名错误")
   	}
   }
   
    ````

   